k2/python/k2/fsa.py

```diff
@@ -59,34 +101,66 @@ def __init__(self,
         self.arcs: RaggedArc = _fsa_from_tensor(tensor)
         self._tensor_attr['score'] = _as_float(self.arcs.values()[:, -1])
         if aux_labels is not None:
-            self.aux_labels = aux_labels
+            self.aux_labels = aux_labels.to(torch.int32)
 
     def _init_internal(self) -> None:
         self._tensor_attr = OrderedDict()
         self._non_tensor_attr = OrderedDict()
 
     def __setattr__(self, name: str, value: Any) -> None:
+        if value is None:
+            return
         if name in ('_tensor_attr', '_non_tensor_attr', 'arcs'):
             object.__setattr__(self, name, value)
         elif isinstance(value, torch.Tensor):
             assert value.shape[0] == self.arcs.values().shape[0]
-            # TODO(fangjun): do we really need a copy here?
+            if name == 'labels':
+                assert value.dtype == torch.int32
+                self.arcs.values()[:, 2] = value
+                return
+
             self._tensor_attr[name] = value.clone()
+
             if name == 'score':
                 assert value.dtype == torch.float32
+                # NOTE: we reinterpret the float patterns
+                # to integer patterns here.
                 self.arcs.values()[:, -1] = _as_int(value.detach())
         else:
             # TODO(fangjun): do we need to copy value?
             self._non_tensor_attr[name] = value
 
     def __getattr__(self, name: str) -> Any:
-        if name in self._tensor_attr:
+        if name == 'labels':
+            return self.arcs.values()[:, 2]
+        elif name in self._tensor_attr:
             return self._tensor_attr[name]
         elif name in self._non_tensor_attr:
             return self._non_tensor_attr[name]
 
         raise AttributeError(f'Unknown attribute {name}')
 
+    def invert_(self) -> 'Fsa':
+        '''Swap the ``labels`` and ``aux_labels``.
+
+        CAUTION:
+          The function name ends with an underscore which means this
+          is an **in-place** operation.
+        '''
+        if hasattr(self, 'aux_labels'):
+            aux_labels = self.aux_labels
+            self.aux_labels = self.labels
+            self.labels = aux_labels
```

Comment:
> A new method invert_() is added for Fsa to swap its labels and aux_labels in-place.

Actually it's not what `Invert` is supposed to do. A naive `Swap` may make the FSA invalid, thinking about the case where the `aux_labels` are sequences (e.g. phones). See function https://github.com/k2-fsa/k2/blob/e9889fa3b202b1fea4b76c370b0f9f0f7efe4c43/k2/csrc/host/aux_labels.h#L149

Reply:
We can give Invert() a different implementation in the case when `aux_labels` has type RaggedTensor(); and we can choose to make Invert_ just crash in that case